### ISSUE TITLE
vscode.Event turns into AnyOf

### ISSUE THREAD

**Issue by Gama11** [haxiomic/dts2hx#29]
This is currently the cause of most compiler errors in vshaxe. The TS definition looks like this:

```ts
/**
 * Represents a typed event.
 *
 * A function that represents an event to which you subscribe by calling it with
 * a listener function as argument.
 *
 * @sample `item.onDidChange(function(event) { console.log("Event happened: " + event); });`
 */
export interface Event<T> {
    /**
     * A function that represents an event to which you subscribe by calling it with
     * a listener function as argument.
     *
     * @param listener The listener function will be called when the event happens.
     * @param thisArgs The `this`-argument which will be used when calling the event listener.
     * @param disposables An array to which a [disposable](#Disposable) will be added.
     * @return A disposable which unsubscribes the event listener.
     */
    (listener: (e: T) => any, thisArgs?: any, disposables?: Disposable[]): Disposable;
}
```

This turn into:

```haxe
typedef Event<T> = ts.AnyOf3<(listener:(e:T) -> Dynamic) -> Disposable, (listener:(e:T) -> Dynamic, thisArgs:Dynamic) -> Disposable, (listener:(e:T) -> Dynamic, thisArgs:Dynamic, disposables:Array<Disposable>) -> Disposable>;
```

It's not totally clear to me why the optional arguments seem to result in an `AnyOf` type, we simply have this:

https://github.com/vshaxe/vscode-extern/blob/cd6c7cc347cf873b0dce732ed1b283bf116079ee/src/vscode/Event.hx

**Comment by haxiomic** [comment]
This is done to better match the function unification behaviour in ts

If you have a signature like
```haxe
function go(callback: (arg1, ?arg2, ?arg3) -> Void)


https://github.com/haxiomic/dts2hx/blob/e4ede65b77740944c20153f5d23d1adadbaf408a/src/ConverterContext.hx#L1266

**Comment by haxiomic** [comment]
What errors does this cause?

**Comment by Gama11** [comment]
Hmm, there's a lot of this, but I guess that's not even really directly related to the AnyOf:

![image](https://user-images.githubusercontent.com/2620907/82739061-e5f72780-9d3c-11ea-87a1-9a208ca4cbc7.png)

But I guess technically our hand-written extern is wrong / shouldn't have `->Void` as the callback return type... The real problem is that there's no `DynamicOrVoidIReallyDontCare` type. :)

**Comment by Gama11** [comment]
This might simply be one of those cases where you still want one of the types to be hand-written and not generated because you have more info how something is actually used in practice (I'm not sure there's much of a use to return something from that callback).

**Comment by haxiomic** [comment]
Yeah, the `DynamicOrVoidIReallyDontCare` type would have come in handy a few times

Maybe keep your old Event.hx and let it override the generated one

**Comment by Gama11** [comment]
Hm, there's another issue though, it seems the generated `Event` isn't really even used in the first place:

![image](https://user-images.githubusercontent.com/2620907/82739207-0c699280-9d3e-11ea-8b2d-dca393313d5d.png)

**Comment by Gama11** [comment]
(which means just keeping that one hand-written Event.hx file doesn't help much)

**Comment by haxiomic** [comment]
That's an intentional change too (for better unification among other things)
https://github.com/haxiomic/dts2hx/blame/e4ede65b77740944c20153f5d23d1adadbaf408a/src/ConverterContext.hx#L1512

onDidChangeVisiblity looks ok to me, what's the haxe error?

**Comment by haxiomic** [comment]
Ohh, I see you need the return there to be void

I'd wrap the callback like `() -> {update(); null;}`

**Comment by Gama11** [comment]
I think we need this in too many places for that to be acceptable. Maybe some solution for "don't care" return is needed on the Haxe side, at least in `haxe.extern`...

**Comment by haxiomic** [comment]
Yeah, I think we'd need a new type in the compiler to allow unification with dynamic or void (`EitherType<Dynamic, Void>` doesn't work fully)

_technically_ I could get it to work by anding more signatures to AnyOf but that would get noisy fast
